### ISSUE TITLE
[BUG] Different indexing behavior for const vs non-const array

### ISSUE THREAD

**Issue by cschreib-ibex** [arrayfire/arrayfire#3073]
Depending on whether an ``af::array`` is ``const`` or not, the same indexing operation produces an array with different dimensions.

Description
===========

* I encountered this bug while refactoring some of my code. The code was a single big function, which I started breaking down into several smaller functions. I carefully marked ``const`` the function parameters that should be constant, and was surprised when this simple refactoring (which shouldn't have changed the output at all) all of a sudden made my program crash.
* Did you build ArrayFire yourself or did you use the official installers: Official installers
* Which backend is experiencing this issue? All
* Do you have a workaround? Yes (don't use ``const``, which isn't great)
* Can the bug be reproduced reliably on your system? Yes
* A clear and concise description of what you expected to happen: I expect that operations on a non-const ``af::array`` form a strict super-set of operations on a const ``af::array``. In other words, if I can do operation ``B`` a const array, I should be able to do the same operation ``B``, with the exact same output, on a non-const array.
* Run your executable with AF_TRACE=all and AF_PRINT_ERRORS=1 environment
  variables set, screenshot or terminal output of the results:
```
[platform][1609772142][007696] [ ..\src\backend\common\DependencyModule.cpp(99) ] Attempting to load: forge.dll
[platform][1609772142][007696] [ ..\src\backend\common\DependencyModule.cpp(102) ] Found: forge.dll
[platform][1609772142][007696] [ ..\src\backend\cuda\device_manager.cpp(428) ] CUDA Driver supports up to CUDA 10.1 ArrayFire CUDA Runtime 10.1
[platform][1609772142][007696] [ ..\src\backend\cuda\device_manager.cpp(495) ] Found 1 CUDA devices
[platform][1609772142][007696] [ ..\src\backend\cuda\device_manager.cpp(521) ] Found device: GeForce GTX 1070 (8 GB | ~6256.71 GFLOPs | 15 SMs)
[platform][1609772142][007696] [ ..\src\backend\cuda\device_manager.cpp(556) ] AF_CUDA_DEFAULT_DEVICE:
[platform][1609772142][007696] [ ..\src\backend\cuda\device_manager.cpp(575) ] Default device: 0(GeForce GTX 1070)
[mem][1609772143][007696] [ ..\src\backend\cuda\memory.cpp(158) ] nativeAlloc:    1 KB 0xb04e00000
```

Reproducible Code and/or Steps
------------------------------
```c++
#include <arrayfire.h>
#include <iostream>

int main() {
    try {
        const std::size_t dim = 10;

        // Create a simple matrix
        af::array m(dim,dim);
        // Keep a constant reference to it
        const af::array& cm = m;

        af::seq i(af::seq(dim), true);
        af::array r1 = m(af::span, i);
        af::array r2 = cm(af::span, i);

        std::cout << "dim r1: " << r1.dims(0) << "," << r1.dims(1) << std::endl; // prints "dim r1: 10,1"
        std::cout << "dim r2: " << r2.dims(0) << "," << r2.dims(1) << std::endl; // prints "dim r2: 10,10" !
    } catch (const af::exception& e) {
        std::cout << e.what() << std::endl;
        return 1;
    }

    return 0;
}
```

Compile the above example with any backend, then notice how the two arrays ``r1`` and ``r2`` have different dimensions, even though they were generated from the almost the same expression. FYI, the original code that triggered this problem was a bit more involved, I stripped it down to its barest bone that reproduces the problem. The original calculation, which makes more sense, is:
```c++
/*const*/ af::array& m = /* input */;
af::array r(m.dims(1));
gfor(af::seq i, static_cast<double>(m.dims(1)))
{
    r(i) = af::sum(m(af::span, i) * m(af::span, i));
}
```
It is effectively ``auto r = af::diag(af::matmulNT(m,m))``, but faster. Happy to learn if there's a better way :)

System Information
------------------
Please provide the following information:
1. ArrayFire version: 3.7.2
2. Devices installed on the system: Nvidia GeForce GTX 1070 (8GB)
3. (optional) Output from the af::info() function if applicable.
```
ArrayFire v3.7.2 (CUDA, 64-bit Windows, build 218dd2c)
Platform: CUDA Runtime 10.1, Driver: 10010
[0] GeForce GTX 1070, 8192 MB, CUDA Compute 6.1
```
4. Output from the following scripts:
Windows: nvidia-smi
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 432.00       Driver Version: 432.00       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1070   WDDM  | 00000000:65:00.0 Off |                  N/A |
|  0%   49C    P8    14W / 151W |    559MiB /  8192MiB |      2%      Default |
+-------------------------------+----------------------+----------------------+
```

Checklist
---------

- [ ] Using the latest available ArrayFire release
- [ ] GPU drivers are up to date

**Comment by 9prady9** [comment]
Is there a specific reason you are setting `gfor` flag to true when you are doing `af::seq i(af::seq(dim), true);` when it is not a `gfor` loop at all.

If you print all the dimensions - it looks like below with `gfor` set to true.
```
dim r1: 10 1 1 10
dim r2: 10 10 1 1
```
That is expected state of dims when gfor is active. In your case, the gfor is set although it is not gfor loop, I believe it was meant to be used for internally setting gfor flag when actually inside gfor loop. @umar456 Can you please confirm if what I said is true or not.

**Comment by cschreib-ibex** [comment]
> Is there a specific reason you are setting gfor flag to true

Yes. The code was originally inside a ``gfor`` loop, I took it apart to only contain the strictly minimal amount of code that reproduces the issue. The original code was something like:
```c++
/*const*/ af::array& m = /* input */;
af::array r(m.dims(1));
gfor(af::seq i, static_cast<double>(m.dims(1)))
{
    r(i) = af::sum(m(af::span, i) * m(af::span, i));
}
```

With a non-const ``m``, it works fine. With a const ``m``, I get a runtime error because of mismatching dimensions in ``operator=``.

**Comment by 9prady9** [comment]
I have to try the gfor based code snippet to see whats going on.

I have tried your standalone code without gfor already. It isn't using `seq` constructor  `af::seq i(af::seq(dim), true);` correctly. The gfor flag argument to seq constructor isn't intended to be used like that.

> Could you clarify why it is expected to get (10,1,1,10) for a non-const variable and (10,10,1,1) for a const variable?

I wasn't saying that is expected behavior w.r.t const-ness of the C++ variable. When you set gfor to true, the code internally moves the batch dimension to largest dimension/axis available. Since that is 4 right now in ArrayFire, the dimensions `[10 10 1 1]` change to `[10 1 1 10]` is expected. That is an implementation detail of gfor and it doesn't alter the state of user `Array` variable. Hence my comment about it.

**Comment by cschreib-ibex** [comment]
Ok my bad. Here's a minimal snippet with gfor that reproduces the problem:
```c++
#include <arrayfire.h>
#include <iostream>

int main() {
    try {
        const std::size_t dim = 10;

        // Create a simple matrix
        af::array m(dim,dim);
        // Keep a constant reference to it
        const af::array& cm = m;

        af::array r(dim);

        // Do simple math with non-const 'm' -- works
        gfor(af::seq i, static_cast<double>(dim)) {
            r(i) = af::sum(m(af::span,i) * m(af::span,i));
        }

        // Do same math with const 'cm' -- fails!
        gfor(af::seq i, static_cast<double>(dim)) {
            r(i) = af::sum(cm(af::span,i) * cm(af::span,i));
        }
    } catch (const af::exception& e) {
        std::cout << e.what() << std::endl;
        return 1;
    }

    return 0;
}
```

Output:
```
ArrayFire Exception (Invalid input size:203):
In function af_err __cdecl af_assign_seq(void **,void *const ,const unsigned int,const struct af_seq *const ,void *const )
In file src\api\c\assign.cpp:183
Invalid dimension for argument 0
Expected: (outDims.ndims() >= inDims.ndims())
 0# af::operator>> in afcuda
 1# af::operator>> in afcuda
 2# af::operator>> in afcuda
 3# af::operator>> in afcuda
 4# main at C:\Users\corentin.schreiber\Documents\aftest.cpp:22
 5# __scrt_common_main_seh at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:253
 6# BaseThreadInitThunk in KERNEL32
 7# RtlUserThreadStart in ntdll

In function class af::array::array_proxy &__cdecl af::array::array_proxy::operator =(const class af::array &)
In file src\api\cpp\array.cpp:564
```

**Comment by cschreib-ibex** [comment]
Confirmed to still be affecting ArrayFire 3.8.0. The above code gives pretty much the same output
```
ArrayFire Exception (Invalid input size:203):
In function af_err __cdecl af_assign_seq(void **,void *const ,const unsigned int,const struct af_seq *const ,void *const )
In file src\api\c\assign.cpp:183
Invalid dimension for argument 0
Expected: (outDims.ndims() >= inDims.ndims())
 0# af::array::array<__half> in afcuda
 1# af::array::array<__half> in afcuda
 2# af::array::array<__half> in afcuda
 3# af::array::array<__half> in afcuda
 4# main at C:\Users\corentin.schreiber\Documents\aftest.cpp:23
 5# __scrt_common_main_seh at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:253
 6# BaseThreadInitThunk in KERNEL32
 7# RtlUserThreadStart in ntdll

In function class af::array::array_proxy &__cdecl af::array::array_proxy::operator =(const class af::array &)
In file src\api\cpp\array.cpp:564
```

**Comment by 9prady9** [comment]
@cschreib-ibex Yes that is correct, I am looking into it already and we will push in a fix most likely in 3.8.1

**Comment by cschreib-ibex** [comment]
Awesome thank you!
